_posts/2025-03-05-#31405.md

```diff
@@ -7,20 +7,46 @@ authors: [mzumsande]
+
+- The `CBlockIndex` objects in this map can be considered the nodes in a tree shaped structure with the genesis block at its root. By definition, each block can only point to a [single predecessor](https://github.com/bitcoin/bitcoin/blob/3c1f72a36700271c7c1293383549c3be29f28edb/src/chain.h#L147), but multiple blocks can point to the same predecessor. Of course, in a single chain, this tree is pruned so that each block will never have more than one block pointing to it. The entire tree structure is kept to enable efficiently handling chain reorgs.
+
+- Block validation is a process that consists of multiple steps, gradually bumping up the `CBlockIndex`'s `nStatus` from its default-initialized `BLOCK_VALID_UNKNOWN` to `BLOCK_VALID_SCRIPTS`. The validation flow which is relevant to this PR can be summarised as:
+	- When the header is successfully validated with `AcceptBlockHeader()`, it is added to `m_block_index`, and generally never removed from it anymore, even if the block fails a subsequent validation step.
+	- When the partial block validation in `AcceptBlock()` succeeds, the block is persisted to disk. This validation is partial, because full validation requires the predecessor block's data, which the node may not have received yet.
+	- As soon as a block header becomes part of the fully validated current chain, `ActivateBestChain()` will attempt to connect the block. If this final validation succeeds, the block is connected and its `nStatus` is [raised](https://github.com/bitcoin/bitcoin/blob/3c1f72a36700271c7c1293383549c3be29f28edb/src/validation.cpp#L2762) to the ultimate `BLOCK_VALID_SCRIPTS`.
+
+- `CBlockIndex` holds a `pprev` pointer to its predecessor which makes it is trivial to iterate backwards (towards the genesis block) over the block tree. Iterating forwards (towards the chain tip) is not possible, because a block can have multiple successors pointing to it. The lack of forward iteration makes operations such as finding a new best block header expensive, because we have to iterate over the entire `m_block_index` block tree again.
+
+- `ChainstateManager::m_best_header` and `ActiveChain.Tip()` are both `CBlockIndex` pointers. During normal operation, they'll often converge to the same block with the highest PoW. However, it is crucial to distinguish their meaning and use. `ActiveChain.Tip()` returns the `CBlockIndex*` of the fully validated and connected block that is the tip of the currently active chain. `m_best_header` points to the header with the most PoW that the node knows about. It represents a node's view of what the most-PoW chain might look like, but it is possible that validation won't allow it to progress there. As such, `m_best_header` is just a hint.
```

Comment:
maybe mention that they are very different during all of IBD.

Reply:
I've added "The distinction is especially visible during IBD, when `ActiveChain().Tip()` progresses towards `m_best_header`, but can only be equal to it when IBD is finished." to the end of this paragraph.